src/UserSpaceInstrumentation/Trampoline.h

```diff
@@ -58,6 +58,58 @@ namespace orbit_user_space_instrumentation {
                                                                     const AddressRange& code_range,
                                                                     uint64_t size);
 
+// Returns the signed 32 bit difference (a-b) between two absolute virtual 64 bit addresses or an
+// error if the difference is too large.
+[[nodiscard]] ErrorMessageOr<int32_t> AddressDifferenceAsInt32(uint64_t a, uint64_t b);
+
+// Merely serves as a return value for the function below.
+struct RelocatedInstruction {
+  // Machine code of the relocated instruction. Might contain multiple instructions to emulate what
+  // the original instruction achieved.
+  std::vector<uint8_t> code;
+
+  // Some relocated instructions contain an absolute address stored in the 'code' above. That
+  // address needs to be adjusted once all the relocations are done. The position of this absolute
+  // address in 'code' is what is stored here.
+  // Example: A conditional jump to a forward position needs to know the
+  // position of an instruction not yet processed.
+  //
+  // Original code does the following: condition cc is true -> InstructionB,
+  // otherwise -> InstructionA, InstructionB
+  //
+  // 0x0100: jcc rip+2 (==0x0104)
+  // 0x0102: InstructionA
+  // 0x0104: InstructionB
+  //
+  // -> relocate ->
+  //
+  // 0x0200: j(!cc) rip+08 (== 0x0210)
+  // 0x0202: jmp [rip+0] (== [0x0208])
+  // 0x0208: 8 byte destination address == address of relocated InstructionB == 0x0217
+  // 0x0210: InstructionA'
+  // 0x0217: InstructionB'
+  //
+  // The conditional jump at 0x0100 is translated into the first three lines of the result. The
+  // address (at 0x0208) of InstructionB' is not yet known at the point of the translation. So it
+  // needs to be recorded and handled later. In this case the `position_of_absolute_address` below
+  // would be 8.
+  std::optional<size_t> position_of_absolute_address = std::nullopt;
```

Comment:
In the case of a conditional jump, there is always only one absolute address required right? That's because if the conditional jump finishes at the boundary of the overwritten 5 bytes, the path where the jump is not taken will automatically continue to the original code. In the case where the relative jump finishes before the 5 bytes boundary, then the following instruction is also in the trampoline, right? Maybe add a comment?

Reply:
I think I don't get the point. 

If you have a conditional jump you get an 8 bit or 32 bit offset. From that you compute the absolute address of the jump target and that is what is written into `code`. And the index of this is stored in `position_of_absolute_address`; it needs to be updated later in case the jump target is also relocated into the trampoline.

Independently from that: If the jump is not taken we just execute the next instruction (as the original conditional jump would do). This is either another relocated instruction or the jump at the end of the trampoline.